app/controllers/application_controller.rb

```diff
@@ -5,6 +5,140 @@ class ApplicationController < Sinatra::Base
+    else
+      redirect '/signup', locals: {message: "Sign up failed, try again."}
+    end
+  end
+
+  get "/login" do
+    if !logged_in?
+      erb :'/users/login'
+    else
+      redirect '/tweets', locals: {message: "You're already logged in!"}
+    end
+  end
+
+  post "/login" do
+    user = User.find_by(username: params[:username])
+    if user && user.authenticate(params[:password])
+      session[:user_id] = user.id
+      redirect "/tweets", locals: {message: "Successfully logged in."}
+    else
+      redirect "/login", locals: {message: "Log in failed, try again."}
+    end
+  end
+
+  get "/logout" do
+    if logged_in?
+      session.clear
+      redirect "/login", locals: {message: "Successfully logged out."}
+    else
+      redirect "/login", locals: {message: "Can't log out, you're not logged in."}
+    end
+  end
+
+  get "/users/:slug" do
+    @user = User.find_by_slug(params[:slug])
+    erb :'/users/show'
+  end
+
+  get "/tweets" do
+    if logged_in?
+      @tweets = Tweet.all
+      erb :'/tweets/tweets'
+    else
+      redirect "/login", locals: {message: "Please log in to see that."}
+    end
+  end
+
+  get "/tweets/new" do
+    if logged_in?
+      erb :'/tweets/new'
+    else
+      redirect "/login", locals: {message: "Please log in to see that."}
+    end
+  end
+
+  post "/tweets" do
+    if logged_in? && params[:content] != ""
+      Tweet.create(content: params[:content], user_id: session[:user_id])
+    elsif params[:content] == ""
+      redirect "/tweets/new", locals: {message: "Tweets can't be empty."}
+    else
+      redirect "/login", locals: {message: "Please log in to see that."}
+    end
+  end
+
+  get "/tweets/:id" do
+    if logged_in?
+      @tweet = Tweet.find(params[:id])
+      erb :'/tweets/show'
+    else
+      redirect "/login", locals: {message: "Please log in to see that."}
+    end
+  end
+
+  get "/tweets/:id/edit" do
+    if logged_in?
+      @tweet = Tweet.find(params[:id])
+      erb :'/tweets/edit'
+    else
+      redirect "/login", locals: {message: "Please log in to see that."}
+    end
+  end
+
+  post "/tweets/:id" do
+    @tweet = Tweet.find(params[:id])
+    if logged_in? && params[:content] != ""
+      @tweet.update(content: params[:content])
+      redirect "/tweets/#{@tweet.id}"
+    elsif params[:content] == ""
+      redirect "/tweets/#{@tweet.id}/edit", locals: {message: "Tweets can't be empty."}
+    else
+      redirect "/login", locals: {message: "Please log in to see that."}
+    end
+  end
+
+  delete "/tweets/:id/delete" do
+    @tweet = Tweet.find(params[:id])
+    if logged_in? && @tweet.user == current_user
+      @tweet.destroy
+      redirect "/tweets"
+    else
+      redirect "/login", locals: {message: "Please log in to see that."}
+    end
+  end
+
+
+  helpers do
+    def current_user
+      User.find(session[:user_id])
+    end
+
+    def logged_in?
+      session[:user_id] ? true : false
```

Comment:
It's better to rely on the successful call to `current_user` than only introspecting on the `session[:id]` value. In theory, a user might have an invalid `id` in their session which would make `logged_in?` return true even though that user would not be findable later. Also, because `current_user` is memoized, it makes sense to call it as early in the request cycle as possible, which generally is the first call `logged_in?`.

```ruby
def logged_in?
  !!current_user
end
```